### PR TITLE
perf: Optimize diff algorithm with insights from #646

### PR DESCRIPTION
**Pull request by yihuang**
it's being running on our testnet, ~~not finished yet~~, but so far so good, the result is identical to previous version.

EDIT: the task previously takes 19hours to run, now takes 14h20m, not bad.

### CODE REVIEW

**Comment by yihuang** [comment]
What do you this about this change, @tac0turtle @cool-develope , the existing round-trip test is pretty exhaustive, and I've done complete run on our testnet data, so the correctness should be pretty solid.

File: diff.go [hunk 1]
```diff
@@ -36,41 +33,105 @@ func (ndb *nodeDB) extractStateChanges(prevVersion int64, prevRoot []byte, root
...
-			newKeys[ibytes.UnsafeBytesToStr(node.key)] = struct{}{}
+
+	var (
+		// current shared node between two versions
+		sharedNode *Node
+		// record all newly added leaf nodes in newer version, it represents all updates and insertions.
+		newLeafNodes []*Node
+		// orphaned leaf nodes in previous version, which represents all deletions and updates.
+		// both `newLeafNodes` and `orphanedLeafNodes` are ordered by key.
+		orphanedLeafNodes []*Node
```
Comment (cool-develope) [inline diff.go:44]: for me, `newLeaves` makes sense than `newLeafNodes`

**Comment by cool-develope** [comment]
I am still worried about the memory usage if we use `extractStateChanges` for the two versions which are far away.
Could we modify it as an iterator or callback way?

File: diff.go (hunk 1 above)
Comment (yihuang) [reply diff.go:44]: sounds good, done.

**Comment by yihuang** [comment]
> I am still worried about the memory usage if we use `extractStateChanges` for the two versions which are far away. Could we modify it as an iterator or callback way?

I'll see what I can do.

**Comment by yihuang** [comment]
> I am still worried about the memory usage if we use `extractStateChanges` for the two versions which are far away. Could we modify it as an iterator or callback way?

done, it's still not strictly constant memory, but I guess it's the best we can do here.

File: diff.go [hunk 2]
```diff
@@ -18,59 +15,146 @@ type KVPair struct {
...
+				Key:   node.key,
+				Value: node.value,
+			}); err != nil {
+				return err
+			}
 		}
-		// skip subtree of shared nodes
-		curIter.Next(shared)
+
+		newLeaves = nil
```
Comment (yihuang) [inline diff.go:58]:
```suggestion
		newLeaves = newLeaves[:0]
```

**Review by cool-develope — APPROVED** [review]
LGTM!

**Comment by yihuang** [comment]
can we backport to 0.19.x @tac0turtle

**Comment by tac0turtle** [comment]
@mergifyio backport release/v0.19.x

**Pull request merged**